src/pv_utils.py

```diff
@@ -0,0 +1,197 @@
+    for name, param_or_buffer in chain(dequantized_model.named_parameters(), dequantized_model.named_buffers()):
+        if name not in master_parameters:
+            continue  # non-quantized weight
+        master_param_or_buffer = master_parameters[name]
+        assert param_or_buffer.shape == master_param_or_buffer.shape
+        unmatched_master_parameters.remove(name)
+    assert len(unmatched_master_parameters) == 0, f"Found unmatched tensors: {unmatched_master_parameters}"
+
+
+def get_original_named_parameters_from_fsdp_module(dequantized_model) -> Dict[str, nn.Parameter]:
+    return {name.replace('_fsdp_wrapped_module.', ''): param for name, param in dequantized_model.named_parameters()}
+
+
+@contextlib.contextmanager
+def print_runtime_stats(operation_name: str, enabled: bool = True, device: Optional[torch.device] = None):
+    if not enabled:
+        yield
+        return
+
+    rank = torch.distributed.get_rank() if torch.distributed.is_initialized() else 0
+    if device is None:
+        device = torch.device(f'cuda:{rank}' if torch.cuda.is_available() else 'cpu')
+    if torch.device.type == 'cuda':
+        torch.cuda.synchronize(device)
+    start_time = time.perf_counter()
+    yield
+    if torch.device.type == 'cuda':
+        torch.cuda.synchronize(device)
+    maybe_distributed_msg = f"rank {rank} " if torch.distributed.is_initialized() else ""
+    print(end=f"{maybe_distributed_msg}{operation_name} took {time.perf_counter() - start_time}\n")
+
+
+def split_quantized_weights_between_ranks(quantized_weights: Dict[str, QuantizedWeight], verify_checksums: bool):
+    """
+    Split all quantized weights between ranks in a distributed setup; uses greedy knapsack heuristic.
+    Note that unlike FSDP, this heuristic will always assign the entire quantized weight to one rank.
+
+    :param quantized_weights: a dictionary [parameter_name] -> QuantizedWeight
+    :returns: a dictionary similar to quantized weights or pointers to different ranks.
+        If your rank stores this quantized weight for [name], then returned_dict[name] is quantized_weights[name]
+        Otherwise, returned_dict[name] = YourQuantizedWeightIsInAnotherRank(rank=where_it_is_stored)
+    :param verify_checksums: if True, synchronize with other ranks and verify that parameters are split consistently.
+        If False, do not synchronize, but instead print a hash of checksum for each rank to be verified by the user.
+    """
+    assert torch.distributed.is_initialized()
+    own_rank = torch.distributed.get_rank()
+    world_size = torch.distributed.get_world_size()
+    all_quantized_weights: Dict[QuantizedWeight, List[str]] = defaultdict(list)
+    for name, quantized_weight in quantized_weights.items():
+        all_quantized_weights[quantized_weight].append(name)
+
+    # order quantized weights in a rank-agnostic way: order by (param size desc, linked param name asc)
+    def _compute_size(qw: QuantizedWeight) -> float:
```

Comment:
smaller scope -> longer name (compute_quantized_weight_size)

Reply:
[chose to keep name over oral discussion]